### DISCUSSION TITLE
[math-utils] should Vector2 math functions modify `this` or return a copy?

### DISCUSSION THREAD

**Discussion by miltoncandelero** [pixijs/pixijs#7553] [RFC]
Among other things, `@pixi/math-utils` will include some vector math operations that will be mixed inside `Point` so you will be able to do something like `myPoint.addition(anotherPoint)` and get the vectorial sum of both Points.

Currently the functions are:

- `addition(other: Point): Point;` & `subtraction(other: Point): Point;`
- `multiplication(other: Point): Point;` & `division(other: Point): Point;`
	- Component to component
- `scalarMultiplication(number: number): Point;`
- `dotProduct(other: Point): number;`
	- https://en.wikipedia.org/wiki/Dot_product
- `crossProduct(other: Point): number;`
	- https://en.wikipedia.org/wiki/Cross_product Cross product doesn't exist on 2d. This is calculated by assuming z=0.
- `normalized(): Point;`
- `magnitude(): number;` & `magnitudeSquared(): number;`
- `project(other: Point): Point;`
	- https://en.wikipedia.org/wiki/Vector_projection
- `reflect(normal: Point): Point;`
	- https://en.wikipedia.org/wiki/Reflection_(mathematics)
- `equals(other: Point): boolean;`
- `distance(other: Point): number;`
	- not really a vector, linear distance between two points.

Some of them have a result of another Point. The issue is:
### **Should these methods modify `this` vector (in-place operation)?**
\- or -
### **Should they always return a `new Point`?**
---
## Precedence:
- Array operations in Javascript sometimes return a new array ([Array.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter)) and sometimes are in place ([Array.sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort))
- [Three.js Vector3 Math ](https://threejs.org/docs/#api/en/math/Vector3)is always in place. You can `clone` first if you need a copy.
- [Babylon.js Vector3 Math ](https://doc.babylonjs.com/typedoc/classes/babylon.vector3)have two versions for each method. One for copy, another for in place.

**Comment by Fxlr8** [comment]
Can we go both? Just pass an optional last argument to every function.

`point.add(other: Point, returnNew?: boolean): Point`

I have seen this in a couple of vector math libs

**Reply by miltoncandelero** [reply]
Instead of a Boolean it can be a reference to the out variable. If left undefined a new one is created and if you want to modify the same point you can pass it as the second parameter.
`myPoint.addition(anotherPoint, outRef)`

**Reply by miltoncandelero** [reply]
Now that I think about it... I think that `toLocal` and `toGlobal` work exactly like that...

**Reply by Fxlr8** [reply]
The `outRef` seems weird. Imagine the situation:
```ts
const pointA = new Point(1, 1)
const pointB = new Point(2, 3)
```

> if you want to modify the same point you can pass it as the second parameter

Do you mean this?
```ts
pointA.addition(pointB, pointA)
```
But this logically equal to this
```ts
pointA.addition(pointB)
```
I think that a.add(b) should modify `a` by default, not create a new object with the result. It seems more intuitive.

**Reply by miltoncandelero** [reply]
It would return a new point unless you give it a reference to use.

I am a sucker for consistency and toLocal and toGlobal already do this...

http://pixijs.download/release/docs/PIXI.DisplayObject.html#toLocal
http://pixijs.download/release/docs/PIXI.DisplayObject.html#toGlobal
